src/sage/groups/libgap_mixin.py

```diff
@@ -987,113 +987,213 @@ def minimum_generating_set(G) -> list:
-    GbyN = phi.ImagesSource()
-    GbyN_mingenset = minimum_generating_set(GbyN)
+        ALGORITHM:
 
-    g = [phi.PreImagesRepresentative(g) for g in GbyN_mingenset]
-    lg = len(g)
+        First, we compute some essential quantities:
 
-    if N.IsAbelian().sage():
-        if is_group_by_gens(G, g):
-            return g
+        `\bold{n} :=\{n_1,n_2\dots n_k\}`
+        where `\{n_1 G_i,n_2G_i \dots n_kG_{i}\}`
+        is any generating set of `G_{i-1}/G_i ,
+        i.e. it's the representative elements of
+        any (prefferably small,
+        but not necessarily minimal)
+        generating set of `G_{i-1}/G_i`
 
-        for i in range(lg):
-            for j in range(len(S)):
-                temp = g[i]
-                g[i] *= S[j]
-                if is_group_by_gens(G, g):
-                    return g
-                g[i] = temp
+        `\bold{N} := \{N_1,N_2\dots N_m\}` where
+        `G_{i-1}/G_i = \{N_1G_i,N_2G_2\dots N_m G_m\}`.
+        This is simply a list of
+        representative elements of `G_{i-1}/G_i`.
 
-        return g + [S[0]]
+        We wish to find the representatives of
+        a minimum generating set of `G/G_i`.
+        Here, we have two cases to consider.
 
-    # A function to generate some combinations of the generators
-    # of the group according to the algorithm. Here it is considered that
-    # the first element of the group N is the identity element.
-    def gen_combinations(g, N, lg):
-        for iter in product(N, repeat=lg):
-            yield [g[i] * iter[lg-i-1] for i in range(lg)]
+        First, if `G_{i-1}/G_i` is abelian :
 
-    N_list = list(N.AsList())
+        if `\braket{gG_i}= G/G_i`, return `g`
 
-    for raw_gens in gen_combinations(g, N_list, lg):
-        if is_group_by_gens(G, raw_gens):
-            return raw_gens
+        for `1 \le p \le s`  and `n_j \in \bold{n}`, we calculate
+        `g^* := \{g_1,g_2\dots g_{p-1} ,g_p n_j,g_{p_1},\dots\}`.
+        If `\braket{g^* G_i} = G/G_i` , return `g^*`
 
-    for raw_gens in gen_combinations(g, N_list, lg):
-        for nl in N_list[1:]:
-            if is_group_by_gens(G, raw_gens+[nl]):
-                return raw_gens + [nl]
+        Second, if `G_{i-1}` is not abelian:
 
-    assert False, "The code should never reach here"
+        First, for all combinations of
+        (not necessarily distinct) elements
+        `N_{i_1},N_{i_2}\dots N_{i_t} \in \bold{N}`,
+        Compute `g^* = \{g_1N_{i_1},g_{i_2}N_{i_3}\dots`
+        `g_{i_t}N_t,g_{t+1}\dots g_s\}`.
+        If `\braket{g^*G_i}\; = G/G_i`, return `{g^*}`
+
+        Then, for all combinations of
+        (not necessarily distinct) elements
+        `N_{i_1},N_{i_2}\dots N_{i_t} N_{i_{t+1}} \in \bold{N}`,
+        Compute `g^* =`
+        `\{g_1N_{i_1},g_{i_2}N_{i_3}\dots`
+        `g_{i_t}N_t,g_{t+1}\dots g_s\}`.
+        If `\braket{g^*G_i}\; = G/G_i`, return `{g^*}`
+
+        By now, we must have exhausted our search.
+
```

Comment:
you should write this whole thing in main docstring of the function i.e. minimum_generating_set , as you can not have a docstring for this nested function and also remove this whole docstring and shift this part to the main function's docstring.